### PR TITLE
feat(clustering) support http forward proxy

### PR DESCRIPTION
**Pull request by fffonion**
### Summary

Allow Dp -> Cp connection to go through a HTTP proxy tunnel. Only HTTP proxy is supported for now.

Things to do after feature frezze before code freeze:

- Make our patch of lua-resty-websocket.client a patch instead of vendoring
- Add support of talk to an HTTP tunnel in https

### Full changelog

* Add forwrd proxy support for clustering data plane

### Issue reference

FTI-2996

### CODE REVIEW

**Comment by fffonion** [comment]
For reviewers:
- patch of websocket.client is done in a seperate commit
- tests is coming

File: kong/clustering/utils.lua
Comment (fffonion) [inline kong/clustering/utils.lua:10]: TODO: use ngx.re.match to replace socket_url, need some unit tests.

File: kong/resty/websocket/client.lua
Comment (flrgh) [inline kong/resty/websocket/client.lua:1]: We now have our own [fork of lua-resty-websocket](https://github.com/Kong/lua-resty-websocket) for any of our extra features that haven't made it upstream yet.

I think we should try to get this change into the fork and then ship it with OSS (currently it's only packaged in EE, but it is trivial to add to OSS). It is currently installed by [kong-build-tools](https://github.com/Kong/kong-build-tools/blob/0f9ecf613a287da09e3f4166ebd71529bf04bf5c/openresty-build-tools/kong-ngx-build#L323-L338) _in-place of_ `resty.websocket`, so any changes/additions just need to be backwards-compatible, but this hasn't proved to be too much of a challenge yet. Feel free to @/ping me for help and/or code review if needed.

If it's not feasible to do make this change before code freeze, would you be willing to do so in a follow-up PR? I'd like us to not have to maintain multiple versions of this library.

File: spec/02-integration/09-hybrid_mode/09-forward-proxy_spec.lua
Comment (flrgh) [inline spec/02-integration/09-hybrid_mode/09-forward-proxy_spec.lua:24]: Argh, it looks like I forgot to port the basic auth support for this fixture back to OSS, so I don't think this is doing anything at the moment :(

If you grab `spec/fixtures/forward-proxy-server.lua` from kong-ee:master it should be fine to drop into this branch as-is.

File: spec/02-integration/09-hybrid_mode/09-forward-proxy_spec.lua
Comment (flrgh) [inline spec/02-integration/09-hybrid_mode/09-forward-proxy_spec.lua:34]: We are removing the legacy hybrid protocol in 3.1 (#9740), so thankfully this extra parameterization can be removed from your integration tests.

**Comment by hbagdi** [comment]
Please make sure to add a [kong.conf.default](https://github.com/Kong/kong/blob/master/kong.conf.default) entry as part of this PR.

File: kong/resty/websocket/client.lua
Comment (fffonion) [reply kong/resty/websocket/client.lua:1]: @flrgh exactly, vendoring this is not a good idea. RIght now i'm vendoring it so I can test it in a single PR (and
it's easier to tell what did we change).
We already have [patch](https://github.com/Kong/kong-build-tools/blob/master/openresty-patches/patches/1.21.4.1/lua-resty-websocket-0.09_01-client-mtls.patch) for lua-resty-websocket, so adding another one won't bring more burden. So yes I will make this a proper patch and remove the vendoring
after feature freeze.

File: spec/02-integration/09-hybrid_mode/09-forward-proxy_spec.lua
Comment (fffonion) [reply spec/02-integration/09-hybrid_mode/09-forward-proxy_spec.lua:24]: i copied it in, and fixed a small typo, will let the ce->ee merge to propogate it to EE

File: kong/tools/utils.lua
Comment (chronolaw) [inline kong/tools/utils.lua:1598]: should localize `string.sub`?

**Comment by chronolaw** [comment]
Should we add `proxy_server` and `cluster_use_proxy` in `spec/fixtures/custom_nginx.template`?

**Comment by fffonion** [comment]
@chronolaw `proxy_server` and `cluster_use_proxy` are kong.conf entries and won't affect the nginx template
(for now).

File: CHANGELOG.md
Comment (chronolaw) [inline CHANGELOG.md:159]: should be `are added`?

File: spec/02-integration/09-hybrid_mode/09-forward-proxy_spec.lua
Comment (chronolaw) [inline spec/02-integration/09-hybrid_mode/09-forward-proxy_spec.lua:2]: style: align `=`

File: spec/01-unit/03-conf_loader_spec.lua
Comment (chronolaw) [inline spec/01-unit/03-conf_loader_spec.lua:972]: Add a blank line before `it(...)`

**Review by flrgh — APPROVED** [review]
:+1: this looks ready to merge

File: spec/01-unit/03-conf_loader_spec.lua
Comment (chronolaw) [inline spec/01-unit/03-conf_loader_spec.lua:1005]: Add a blank line before `it(...)`

File: spec/01-unit/03-conf_loader_spec.lua
Comment (chronolaw) [inline spec/01-unit/03-conf_loader_spec.lua:1035]: Add a blank line after `it(...)`

File: spec/02-integration/09-hybrid_mode/09-forward-proxy_spec.lua
Comment (chronolaw) [inline spec/02-integration/09-hybrid_mode/09-forward-proxy_spec.lua:51]: Now we can remove `legacy_hybrid_protocol`

File: spec/02-integration/09-hybrid_mode/09-forward-proxy_spec.lua
Comment (chronolaw) [inline spec/02-integration/09-hybrid_mode/09-forward-proxy_spec.lua:62]: remove `legacy_hybrid_protocol`

File: CHANGELOG.md
Comment (dndx) [inline CHANGELOG.md:158]:
```suggestion
- Add HTTP CONNECT forward proxy support for Hybrid Mode connections. New configuration
```

Make it explicit we only support `HTTP CONNECT` proxy at this moment.

File: kong-3.1.0-0.rockspec
Comment (dndx) [inline kong-3.1.0-0.rockspec:97]: We need to move this into a patch file, please do not copy this entire file into the Kong repo.

File: kong.conf.default
Comment (dndx) [inline kong.conf.default:225]:
```suggestion
#proxy_server =                  # Proxy server defined as a URL. Kong will only use
```

File: kong.conf.default
Comment (dndx) [inline kong.conf.default:400]:
```suggestion
                         # for Hybrid mode connections if this option is turned on.
```

File: kong/clustering/utils.lua
Comment (dndx) [inline kong/clustering/utils.lua:333]: This won't work with HTTPS support. Just a note.

File: kong/clustering/utils.lua
Comment (dndx) [inline kong/clustering/utils.lua:425]: Will this URL contains the credentials?

File: kong/conf_loader/init.lua
Comment (dndx) [inline kong/conf_loader/init.lua:1005]:
```suggestion
      errors[#errors + 1] = "fragments, query strings or parameters are meaningless in proxy configuration"
```

File: kong/tools/utils.lua
Comment (dndx) [inline kong/tools/utils.lua:1590]: Can't we use https://github.com/ledgetech/lua-resty-http/blob/master/lib/resty/http.lua#L255 instead of copying?

File: kong-3.1.0-0.rockspec
Comment (chronolaw) [reply kong-3.1.0-0.rockspec:97]: I think it will be done in other PR after this merged.

File: kong/tools/utils.lua
Comment (fffonion) [reply kong/tools/utils.lua:1590]: that one doesn't support credential inside URI, it will need change. instead of bringing uncertainty,
i would suggest go with known-to-be good solution.

File: kong/clustering/utils.lua
Comment (fffonion) [reply kong/clustering/utils.lua:425]: no it will be the scheme://host:port only

File: kong.conf.default
Comment (fffonion) [reply kong.conf.default:225]: leaving a trailing `=` follows the pattern in kong.conf.default, consider this resolved

File: kong/clustering/utils.lua
Comment (fffonion) [reply kong/clustering/utils.lua:333]: good catch, move default to 443

File: kong.conf.default
Comment (dndx) [reply kong.conf.default:225]: This is what I found: 

https://github.com/Kong/kong/blob/3143aa154d7d7bdf1476f13cb0136527685e259e/kong.conf.default#L268-L318
https://github.com/Kong/kong/blob/3143aa154d7d7bdf1476f13cb0136527685e259e/kong.conf.default#L658-L662

File: kong.conf.default
Comment (dndx) [inline kong.conf.default:225]:
```suggestion
#proxy_server =                  # Proxy server defined as a URL. Kong will only use this
```

File: kong/clustering/utils.lua
Comment (dndx) [inline kong/clustering/utils.lua:12]:
```suggestion
local parse_url = require("socket.url").parse
```

To follow the style of existing code in this file.

File: kong/clustering/utils.lua
Comment (dndx) [inline kong/clustering/utils.lua:330]:
```suggestion


```

**Review by dndx — APPROVED** [review]
Approved with minor style recommendation, and can be merged after they have been addressed and tests are green.

We need to remove the copied code from lua-resty-webaocket still, but it is decided to do that in a separate PR.

**Pull request merged**
